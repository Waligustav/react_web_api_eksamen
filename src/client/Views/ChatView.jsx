import React, { useEffect, useState } from "react";

export function ChatView({ chatPreview, username, onSendMessage }) {
  const [chatLog, setChatLog] = useState([]);
  const [message, setMessage] = useState("");
  const [ws, setWs] = new useState();

  useEffect(() => {
    const ws = new WebSocket("ws://localhost:3000");
    ws.onmessage = (event) => {
      const { message, id, username } = JSON.parse(event.data);
      setChatLog((chatLog) => {
        return [...chatLog, { message, id, username }];
      });
    };
    ws.onopen = (event) => {
      ws.send(
        JSON.stringify({
          type: "login",
          username,
        })
      );
    };
    setWs(ws);
  }, []);

  function handleSubmitMessage(e) {
    e.preventDefault();
    if (ws.readyState !== 0) {
      ws.send(
        JSON.stringify({
          type: "message",
          message: message,
        })
      );
    }
    setMessage("");
  }

  const actualChat = chatLog.length ? chatLog : chatPreview;
  return (
    <main className="chatContainer">
      <header>
        <h1>Welcome to the chat</h1>
        <h3>Type something and hit "send"</h3>
      </header>
      <main className="chatLog">
        {actualChat?.map(({ message, id, username }) => (
          <div className="message" key={id}>
            <strong>{username}: </strong>
            {message}
          </div>
        ))}
      </main>
      <footer>
        <form onSubmit={handleSubmitMessage}>
          <input
            type="text"
            autoFocus={true}
            value={message}
            onChange={(e) => setMessage(e.target.value)}
          />
          <button>Send</button>
        </form>
      </footer>
    </main>
  );
}
