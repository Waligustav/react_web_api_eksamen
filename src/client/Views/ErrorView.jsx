import React from "react";

export function ErrorView({ error }) {
  return (
    <div className="mainContainerLogin">
      <h1>An error has occurred</h1>
      <div>{error.toString()}</div>
    </div>
  );
}
