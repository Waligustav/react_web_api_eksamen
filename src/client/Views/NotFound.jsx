import React from "react";

export const NotFound = () => {
  return (
    <div>
      <p>Error: Page not found</p>
    </div>
  );
};
