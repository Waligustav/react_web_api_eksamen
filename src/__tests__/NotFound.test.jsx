import { NotFound } from "../client/Views/NotFound";
import React from "react";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";

describe("shows 'not found' view", () => {
  it("dom displays 'not found' view", () => {
    const container = document.createElement("div");
    document.body.appendChild(container);
    act(() => {
      ReactDOM.render(<NotFound />, container);
    });
    expect(container.innerHTML).toMatchSnapshot();
  });
});
