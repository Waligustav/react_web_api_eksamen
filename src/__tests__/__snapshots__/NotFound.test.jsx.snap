// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows 'not found' view dom displays 'not found' view 1`] = `"<div><p>Error: Page not found</p></div>"`;
